File: packages/shared/src/scoring/score-label.ts
/**
 * Score labels are returned by the API as stable codes and translated in the
 * frontend, so the same rating reads correctly in every supported locale.
 */
export const SCORE_LABELS = [
  'NO_REVIEWS',
  'FEW_REVIEWS',
  'OVERWHELMINGLY_POSITIVE',
  'VERY_POSITIVE',
  'POSITIVE',
  'MOSTLY_POSITIVE',
  'MIXED',
  'MOSTLY_NEGATIVE',
  'NEGATIVE',
  'VERY_NEGATIVE',
] as const;

export type ScoreLabel = (typeof SCORE_LABELS)[number];

/** Sentiment bucket used for colour coding in the UI. */
export type ScoreSentiment = 'positive' | 'mixed' | 'negative' | 'unknown';

/**
 * Below this many reviews a game gets the neutral `FEW_REVIEWS` label instead
 * of a confident one. A single glowing review must never read as "acclaimed".
 */
export const DEFAULT_SCORE_LABEL_MINIMUM_REVIEWS = 10;

/**
 * The strongest label additionally requires a large sample, so it stays a
 * meaningful distinction rather than something 12 friendly reviews can buy.
 */
export const OVERWHELMING_LABEL_MINIMUM_REVIEWS = 200;

/**
 * Resolves the display label for a game from its positive percentage and the
 * size of its sample.
 *
 * @param positivePercentage positive share of reviews, 0..100
 * @param totalReviews number of reviews backing that percentage
 */
export function resolveScoreLabel(
  positivePercentage: number,
  totalReviews: number,
  minimumReviews: number = DEFAULT_SCORE_LABEL_MINIMUM_REVIEWS,
): ScoreLabel {
  if (totalReviews <= 0) {
    return 'NO_REVIEWS';
  }
  if (totalReviews < minimumReviews) {
    return 'FEW_REVIEWS';
  }
  if (positivePercentage >= 95 && totalReviews >= OVERWHELMING_LABEL_MINIMUM_REVIEWS) {
    return 'OVERWHELMINGLY_POSITIVE';
  }
  if (positivePercentage >= 85) {
    return 'VERY_POSITIVE';
  }
  if (positivePercentage >= 75) {
    return 'POSITIVE';
  }
  if (positivePercentage >= 60) {
    return 'MOSTLY_POSITIVE';
  }
  if (positivePercentage >= 45) {
    return 'MIXED';
  }
  if (positivePercentage >= 30) {
    return 'MOSTLY_NEGATIVE';
  }
  if (positivePercentage >= 15) {
    return 'NEGATIVE';
  }
  return 'VERY_NEGATIVE';
}

export function resolveScoreSentiment(label: ScoreLabel): ScoreSentiment {
  switch (label) {
    case 'OVERWHELMINGLY_POSITIVE':
    case 'VERY_POSITIVE':
    case 'POSITIVE':
    case 'MOSTLY_POSITIVE':
      return 'positive';
    case 'MIXED':
      return 'mixed';
    case 'MOSTLY_NEGATIVE':
    case 'NEGATIVE':
    case 'VERY_NEGATIVE':
      return 'negative';
    default:
      return 'unknown';
  }
}

File: apps/api/src/GameScore.Domain/Scoring/ScoreLabel.cs
namespace GameScore.Domain.Scoring;

public static class ScoreLabels
{
    public const string NoReviews = "NO_REVIEWS";
    public const string FewReviews = "FEW_REVIEWS";
    public const string OverwhelminglyPositive = "OVERWHELMINGLY_POSITIVE";
    public const string VeryPositive = "VERY_POSITIVE";
    public const string Positive = "POSITIVE";
    public const string MostlyPositive = "MOSTLY_POSITIVE";
    public const string Mixed = "MIXED";
    public const string MostlyNegative = "MOSTLY_NEGATIVE";
    public const string Negative = "NEGATIVE";
    public const string VeryNegative = "VERY_NEGATIVE";

    public const int DefaultMinimumReviews = 10;
    public const int OverwhelmingMinimumReviews = 200;

    public static string Resolve(
        double positivePercentage,
        int totalReviews,
        int minimumReviews = DefaultMinimumReviews)
    {
        if (totalReviews <= 0)
        {
            return NoReviews;
        }

        if (totalReviews < minimumReviews)
        {
            return FewReviews;
        }

        if (positivePercentage >= 95 && totalReviews >= OverwhelmingMinimumReviews)
        {
            return OverwhelminglyPositive;
        }

        if (positivePercentage >= 85)
        {
            return VeryPositive;
        }

        if (positivePercentage >= 75)
        {
            return Positive;
        }

        if (positivePercentage >= 60)
        {
            return MostlyPositive;
        }

        if (positivePercentage >= 45)
        {
            return Mixed;
        }

        if (positivePercentage >= 30)
        {
            return MostlyNegative;
        }

        if (positivePercentage >= 15)
        {
            return Negative;
        }

        return VeryNegative;
    }

    public static string ResolveSentiment(string label) => label switch
    {
        OverwhelminglyPositive or VeryPositive or Positive or MostlyPositive => "positive",
        Mixed => "mixed",
        MostlyNegative or Negative or VeryNegative => "negative",
        _ => "unknown",
    };
}
